lib/imagemounter/utils/list_developer_image.js

```diff
@@ -0,0 +1,184 @@
+ * @property {string} sha
+ * @property {number} size
+ * @property {string} url
+ */
+
+/**
+ * @typedef {Object} GithubTreeResponse
+ * @property {string} sha
+ * @property {string} url
+ * @property {GithubTreeObject[]?} tree
+ * @property {boolean?} truncated
+ * @property {string?} node_id
+ * @property {number?} size
+ * @property {string?} content
+ * @property {string?} base64
+ */
+
+const AVAILABLE_REPOS = [
+    {
+        user: 'JinjunHan',
+        name: 'iOSDeviceSupport',
+        subFolderList: ['iOSDeviceSupport']
+    }, {
+        user: 'filsv',
+        name: 'iOSDeviceSupport',
+        subFolderList: []
+    }, {
+        user: 'iGhibli',
+        name: 'iOS-DeviceSupport',
+        subFolderList: ['DeviceSupport']
+    }
+];
+const DEFAULT_IMAGE_DIR_NAME = 'iOSDeviceSupport';
+const DEVELOPER_IMAGE_FILE_NAME = 'DeveloperDiskImage.dmg';
+const DEVELOPER_IMAGE_SIGNATURE_FILE_NAME = 'DeveloperDiskImage.dmg.signature';
+
+/**
+ * Use list api to return the file list of folder.
+ * @param {string} user github user
+ * @param {string} name the name of repository
+ * @param {string[]} subFolderList subfolder list in level order
+ * @returns {GithubTreeObject[]?} file list under target folder
+ */
+async function listGithubImageList(user, name, subFolderList = []) {
+    const initialUrl = `https://api.github.com/repos/${user}/${name}/git/trees/master`;
+    const repoUrl = `https://github.com/${user}/${name}/`;
+    const fileList = [];
+    let curUrl = initialUrl;
+    /**
+     * @type {GithubTreeObject[]?}
+     */
+    let ret = null;
+    for (let i = 0; i <= subFolderList.length; i++) {
+        const res = await axios.get(curUrl);
+        /**
+         * @type {GithubTreeResponse}
+         */
+        const body = res.data;
+        const treeList = body?.tree;
+        if (!treeList) {
+            throw new Error(`Failed on looking up ${fileList.join('/')} under ${repoUrl}: ${JSON.stringify(body)}`);
+        }
+        if (i === subFolderList.length) {
+            ret = treeList;
+            break;
+        }
+        const entry = subFolderList[i];
+        const nextItem = _.find(treeList, (item) => item.path === entry);
+        if (!nextItem) {
+            throw new Error(`Unable to find ${entry} under ${fileList.join('/')} in under ${repoUrl}: ${JSON.stringify(treeList)}`);
+        }
+        if (nextItem.type !== 'tree') {
+            throw new Error(`${entry} under ${fileList.join('/')} in ${repoUrl} is expected to be a tree, got ${nextItem.type} instead.`);
+        }
+        fileList.push(entry);
+        curUrl = nextItem.url;
+    }
+    return ret;
+}
+
+/**
+ * Find `DeveloperDiskImage.dmg` recursively under folder and subfolder.
+ * @param {string} entry current folder
+ * @returns {string | undefined} parent folder of `DeveloperDiskImage.dmg` or `undefined` if no such file exists
+ */
+async function findDeveloperImageFromDirectory(entry) {
+    const fileList = await readdir(entry);
```

Comment:
If you use option `{withFileTypes: true}` for `readdir()`, you will get an object that has an `isFile()` and `isDirectory()` method, so you won't need to `lstat` unless it's a symbolic link.  see [docs](https://nodejs.org/api/fs.html#fsreaddirpath-options-callback)

Reply:
New skill fetched!